Cache supported tags in resolver factory

ichard26 opened the pull request:
I've observed get_supported() consume up to 10% of the installation step when installing many packages. Each call can take 1-5ms (presumably only on Linux due to the large number of supported tags) and there is one call for every lookup in the cache. As all of these calls are made with no arguments, the tags can be trivially queried in advance during factory initialization.

<details><summary>Example profile</summary>
<p>

`python -m cProfile -o profile.pstats -m pip install --only-binary=:all: --no-index --ignore-installed --no-deps --find-links homeassistant/deps/ -r homeassistant/requirements.txt` (from https://github.com/home-assistant/core/blob/2024.6.0/requirements.txt)

`get_supported()` is left of `install_given_reqs()` which is coloured green.

![pstats](https://github.com/pypa/pip/assets/63936253/4de31d07-1025-421f-a372-449a8a20371b)

</p>
</details> 

Towards #12712.

uranusjr reviewed (approved):
Seems like a quick win.

morotti commented:
I really hope this patch get merged soon. ❤️ ❤️ ❤️ 

I am getting ~17% faster `pip install ...` on my machine and my venvs.
The tag calculations take as long as it takes to download and resolve all packages!

ichard26 commented:
Rebased on main and updated NEWS entry to more accurately reflect that this improves dependency resolution performance, not exclusively install performance. If CI stays :green_apple:, I'd like to see this merged.

pfmoore commented:
Nice improvement, thanks!
